### PR TITLE
Permutation iterator tests

### PR DESCRIPTION
**Pull request by SergeyKopienko**

### CODE REVIEW

File: test/parallel_api/algorithm/alg.modifying.operations/transform_permutation.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    }
+}
+
+int
+main()
+{
+    test<std::int32_t, std::int32_t>();
+    test<std::int32_t, float32_t>();
+    test<std::uint16_t, float32_t>();
+    test<float32_t, float64_t>();
```
Comment (MikeDvorskiy) [inline test/parallel_api/algorithm/alg.modifying.operations/transform_permutation.pass.cpp:106]: I think `permuation_iterator ` based on floating number index doesnt make sense.

Comment (SergeyKopienko) [reply test/parallel_api/algorithm/alg.modifying.operations/transform_permutation.pass.cpp]: Removed.

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 2]
```diff
@@ -39,13 +81,490 @@ main()
...
+{
+    DEFINE_TEST_CONSTRUCTOR(test_sort)
+
+    template <typename TIterator>
+    void generate_data(TIterator itBegin, TIterator itEnd)
+    {
+        const auto testing_n = ::std::distance(itBegin, itEnd);
+
+        for (auto it = itBegin; it != itEnd; ++it)
+            *it = testing_n - std::distance(itBegin, it);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:106]: probably it make sense to add just integer counter instead of distance calculation every iteration within the loop?

File: test/support/utils_invoke.h [hunk 3]
```diff
@@ -253,6 +253,47 @@ struct invoke_on_all_policies
...
+    }
+
+    template <typename TStepFunctor, typename Operand>
+    void
+    operator()(TStepFunctor stepOp, Operand op)
+    {
+        for (TSourceDataSize perm_idx_step = 1; perm_idx_step < src_data_size; perm_idx_step = stepOp(perm_idx_step))
+        {
+            TSourceDataSize src_data_idx = 0;
+            for (TSourceDataSize perm_idx = 0; perm_idx < src_data_size; perm_idx += perm_idx_step, ++src_data_idx)
```
Comment (MikeDvorskiy) [inline test/support/utils_invoke.h:278]: Probably, it makes sense to write it simpler? without empty loop?
```
TSourceDataSize src_data_idx = src_data_size;
TSourceDataSize perm_idx = perm_idx_step*src_data_size;

```

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 4]
```diff
@@ -39,13 +81,490 @@ main()
...
+{
+    DEFINE_TEST_CONSTRUCTOR(test_partial_sort)
+
+    template <typename TIterator>
+    void generate_data(TIterator itBegin, TIterator itEnd)
+    {
+        const auto testing_n = ::std::distance(itBegin, itEnd);
+
+        for (auto it = itBegin; it != itEnd; ++it)
+            *it = testing_n - std::distance(itBegin, it);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:154]: Please see a comment  above about std::distance.

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 5]
```diff
@@ -39,13 +81,490 @@ main()
...
+// dpl::reduce, dpl::transform_reduce -> __parallel_transform_reduce
+// Requirements: only for random_access_iterator
+DEFINE_TEST(test_transform_reduce)
+{
+    DEFINE_TEST_CONSTRUCTOR(test_transform_reduce)
+
+    template <typename TIterator>
+    void generate_data(TIterator itBegin, TIterator itEnd)
+    {
+        const auto testing_n = ::std::distance(itBegin, itEnd);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:331]: unusable variable

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 6]
```diff
@@ -39,13 +81,490 @@ main()
...
+    
+        TestValueType val;
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:333]: uninitialized variable

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 7]
```diff
@@ -39,13 +81,490 @@ main()
...
+
+    template <typename Policy, typename Iterator1, typename Iterator2, typename Size>
+    ::std::enable_if_t<is_random_access_iterator_v<Iterator1>>
+    operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Iterator2 first2, Iterator1 last2, Size n)
+    {
+        TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, n);     // source data(1) for merge
+        TestDataTransfer<UDTKind::eVals, Size> host_vals(*this, n);     // source data(2) for merge
+        
+        // Fill full source data set
+        generate_data(host_keys.get(), host_keys.get() + n, TestValueType{});
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:392]: `std::iota(host_keys.get(), host_keys.get() + n, TestValueType{})` ?

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 8]
```diff
@@ -39,13 +81,490 @@ main()
...
+        generate_data(host_vals.get(), host_vals.get() + n, TestValueType{} + n / 2);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:393]: `std::iota(host_vals.get(), host_vals.get() + n, TestValueType{} + n / 2) ?`

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 9]
```diff
@@ -39,13 +81,490 @@ main()
...
+
+        test_through_permutation_iterator_mibt{first1, n}(
+            [&](auto permItBegin1, auto permItEnd1)
+            {
+                test_through_permutation_iterator_mibt{first2, n}(
+                    [&](auto permItBegin2, auto permItEnd2)
+                    {
+                        std::vector<TestValueType> resultBuf;
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:401]: `std::vector<TestValueType> resultBuf(2 * n);` instead of two lines?

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 10]
```diff
@@ -39,13 +81,490 @@ main()
...
+                        resultBuf.resize(2 * n);
+
+                        const auto resultEnd = dpl::merge(exec, permItBegin1, permItEnd1, permItBegin2, permItEnd2, resultBuf.begin());
+                        // TODO wait_and_throw(exec);
+
+                        std::vector<TestValueType> expected;
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:407]: `std::vector<TestValueType> expected(3 * n)`; instead of two lines?

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 11]
```diff
@@ -39,13 +81,490 @@ main()
...
+                        expected.resize(3 * n);
+
+                        auto expectedEnd = std::merge(permItBegin1, permItEnd1, permItBegin2, permItEnd2, expected.begin());
+
+                        const auto expectedSize = ::std::distance(expected.begin(), expectedEnd);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:412]: Due yo we have `::std::enable_if_t<is_random_access_iterator_v<Iterator1>>`
I think we can write just `-` instead of `std::distance`.

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 12]
```diff
@@ -39,13 +81,490 @@ main()
...
+    }
+
+    template <typename Policy, typename Iterator1, typename Size>
+    ::std::enable_if_t<is_random_access_iterator_v<Iterator1>>
+    operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Size n)
+    {
+        TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, n);     // source data for find
+        
+        // Fill full source data set
+        generate_data(host_keys.get(), host_keys.get() + n);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:447]: `std::iota(host_keys.get(), host_keys.get() + n, 0);`

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 13]
```diff
@@ -39,13 +81,490 @@ main()
...
+    template <typename Policy, typename Iterator1, typename Size>
+    ::std::enable_if_t<is_random_access_iterator_v<Iterator1>>
+    operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Size n)
+    {
+        for (bool bCallMakeHeap : {false, true})
+        {
+            TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, n);     // source data for find
+        
+            // Fill full source data set
+            generate_data(host_keys.get(), host_keys.get() + n);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:496]: `std::iota(host_keys.get(), host_keys.get() + n, 0);` ?

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 14]
```diff
@@ -39,13 +81,490 @@ main()
...
+
+template <::std::size_t CallNumber>
+using invoke_on_all_hetero_policies_wrapper = invoke_on_all_hetero_policies<CallNumber>;
+
+#endif // TEST_DPCPP_BACKEND_PRESENT
+
+int
+main()
+{
+#if TEST_DPCPP_BACKEND_PRESENT
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:560]: The following code block  confuses me....
`#if TEST_DPCPP_BACKEND_PRESENT` tells that code block is indented just for DPPP backend (hetero backend).
But on the line 564 we call the tests with the host policy - `invoke_on_all_host_policies_wrapper`.
and on the line 567 we dont call the tests with hetero policy (?)

**Comment by MikeDvorskiy** [comment]
@SergeyKopienko could you please add a description to this PR - which test cases does PR cover?
1)....
2)....
3)...
etc

File: test/parallel_api/iterator/permutation_iterator.pass.cpp (hunk 14 above)
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp:562]: Fixed

File: test/parallel_api/iterator/permutation_iterator.pass.cpp (hunk 2 above)
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp]: Fixed

File: test/parallel_api/iterator/permutation_iterator.pass.cpp (hunk 4 above)
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp]: Fixed

File: test/parallel_api/iterator/permutation_iterator.pass.cpp (hunk 12 above)
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp]: Fixed

File: test/parallel_api/iterator/permutation_iterator.pass.cpp (hunk 13 above)
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp]: Fixed

File: test/parallel_api/iterator/permutation_iterator.pass.cpp (hunk 11 above)
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp:398]: As described at https://en.cppreference.com/w/cpp/iterator/distance. there are specialization for this case.
An I prefer `std::distance`

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 15]
```diff
@@ -39,13 +81,490 @@ main()
...
+{
+    DEFINE_TEST_CONSTRUCTOR(test_transform_reduce)
+
+    template <typename TIterator>
+    void generate_data(TIterator itBegin, TIterator itEnd)
+    {
+        const auto testing_n = ::std::distance(itBegin, itEnd);
+    
+        TestValueType val;
+        for (auto it = itBegin; it != itEnd; ++it)
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp]: Done

File: test/support/utils_invoke.h (hunk 3 above)
Comment (SergeyKopienko) [reply test/support/utils_invoke.h]: Fixed

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 16]
```diff
@@ -1,8 +1,17 @@
-//==============================================================
-// Copyright © 2020 Intel Corporation
+// -*- C++ -*-
+//===-- binary_search.pass.cpp --------------------------------------------===//
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:2]: It seems that something wrong with "file header" comment....

File: test/parallel_api/iterator/permutation_iterator.h [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+    void
+    operator()(TStepFunctor stepOp, Operand op)
+    {
+        ::std::vector<TSourceDataSize> indexes;
+
+        for (TSourceDataSize perm_idx_step = 1; perm_idx_step < data.src_data_size; perm_idx_step = stepOp(perm_idx_step))
+        {
+            indexes.clear();
+            for (TSourceDataSize perm_idx = 0; perm_idx < data.src_data_size; perm_idx += perm_idx_step)
+                indexes.push_back(perm_idx);
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.h:115]: To avoid deallocation/allocation/copiyng overhead on std::vector::push_back you should reserve capacity for the vector firstly.
For example, see cpprefererence 
[std::vector](http://en.cppreference.com/w/cpp/container/vector)<int, NAlloc<int>> v1;
```
        v1.reserve(max_elements); // reserves at least max_elements * sizeof(int) bytes
 
        for (int n = 0; n < max_elements; ++n)
            v1.push_back(n);
```

File: test/parallel_api/iterator/permutation_iterator.h [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+
+        for (TSourceDataSize perm_idx_step = 1; perm_idx_step < data.src_data_size; perm_idx_step = stepOp(perm_idx_step))
+        {
+            TSourceDataSize* itIndexStart = test_base_data.get_start_from(TestUtils::UDTKind::eKeys);
+            TSourceDataSize* itIndex = itIndexStart;
+            TSourceDataSize permItCount = 0;
+            for (TSourceDataSize perm_idx = 0; perm_idx < data.src_data_size; perm_idx += perm_idx_step, ++permItCount)
+            {
+                *itIndex = perm_idx;
+                ++itIndex;
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.h:151]: It looks that `itIndex` is not used below.

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 19]
```diff
@@ -39,15 +155,557 @@ main()
     size_t num_elements = std::distance(first, last) / 2 + std::distance(first, last) % 2;
     auto permutation_first = dpl::make_permutation_iterator(first, multiply_index_by_two());
     auto permutation_last = permutation_first + num_elements;
+
+    //auto it = std::copy(dpl::execution::make_device_policy(q), permutation_first, permutation_last, result.begin());
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:159]: forgot to remove?

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 20]
```diff
@@ -39,15 +155,557 @@ main()
...
+    void
+    operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Iterator2 first2, Iterator2 last2, Iterator3 first3, Iterator3 last3, Size n)
+    {
+        if constexpr (can_run_nonmodify_test<Iterator1>())
+        {
+            TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, n);     // source data(1) for merge
+            TestDataTransfer<UDTKind::eVals, Size> host_vals(*this, n);     // source data(2) for merge
+            TestDataTransfer<UDTKind::eRes,  Size> host_res (*this, n);     // merge results
+
+            if (::std::distance(first3, last3) < (::std::distance(first1, last1) + ::std::distance(first2, last2)))
```
Comment (MikeDvorskiy) [inline test/parallel_api/iterator/permutation_iterator.pass.cpp:475]: It is better write as following
`assert(::std::distance(first3, last3) >= (::std::distance(first1, last1) + ::std::distance(first2, last2)))`

File: test/parallel_api/iterator/permutation_iterator.pass.cpp [hunk 21]
```diff
@@ -39,15 +155,557 @@ main()
...
+
+            assert(n > 0);
+            const auto valueToFind = *host_keys.get();
+
+            test_through_permutation_iterator_mibt<Iterator1, Size, PermItIndexTag>{first1, n}(
+                [&](auto permItBegin, auto permItEnd)
+                {
+                    if (2 <= ::std::distance(permItBegin, permItEnd))
+                    {
+                        //const auto valueToFind = *(permItBegin + 1);
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/parallel_api/iterator/permutation_iterator.pass.cpp]: Fixed

**Comment by MikeDvorskiy** [comment]
There many CI fails....
Especially some runtime error with memory:

```
02/172 Test #102: permutation_iterator.pass ...................Subprocess aborted***Exception:   0.16 sec
[3203](https://github.com/oneapi-src/oneDPL/actions/runs/7128854244/job/19411637682?pr=1293#step:6:3204)double free or corruption (out) 
```
([HOST,bknd=serial,cmplr=g++,ubuntu-20.04,std=с++17,cfg=release](https://github.com/oneapi-src/oneDPL/actions/runs/7128854244/job/19411640715?pr=1293#logs))

**Comment by SergeyKopienko** [comment]
@MikeDvorskiy, @danhoeflinger
FYI: there is 3 temporary commits with "@@@" in the start of message which I am not propose to merge into main branch.

**Comment by SergeyKopienko** [comment]
@MikeDvorskiy, @danhoeflinger
Looks like Ci is going to finish with GREEN state (excepting two tasks)

**Review by MikeDvorskiy — DISMISSED** [review]
LGTM

**Review by MikeDvorskiy — DISMISSED** [review]
LGTM

**Review by danhoeflinger — COMMENTED** [review]
I've only really looked at this at a high level, not a very detailed review.
I'm happy you've put this together and It expands the coverage a lot for permutation iterator, and helps to test a very complicated system, especially with #1254 under review.

If I am correct in my understanding, this adds coverage to use permutation with a variety of index map iterator types (counting, host, USM, etc. ...) within a variety of other oneDPL algorithms.  This is good, but I believe we are always using similar simple base iterators for the permutation iterator.  This is another axis of complexity which is not covered by this PR.  

In other words, for `onedpl_alg( permutation_iterator(base_iter, map_iter))`,  we are expanding coverage by varying `onedpl_alg`, and `map_iter`, but not by varying `base_iter`.  

This isn't a blocker for this PR, but may still be a blindspot with permutation_iterator despite these additions.  `transform_iterator`, `zip_iterator` are both cases I'm interested in.  (It could also be interesting to test `transform_iterator` in the `map_iter`).

I know some of these cases I'm describing might be officially "unsupported", or a "known limitation".

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 22]
```diff
@@ -460,8 +460,7 @@ struct __get_sycl_range
         auto __n = __last - __first;
         assert(__n > 0);
 
-        auto res_src = this->operator()(__first.base(),
-                                        __first.base() + __dpl_sycl::__get_buffer_size(__first.base().get_buffer()));
+        auto res_src = this->operator()(__first.base(), oneapi::dpl::end(__first.base().get_buffer()));
```
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:463]: It is a bugfix.
There is a separate commit "[oneDPL][sycl] + taking  into account sycl_iterator 'offset' parameter within specialization for permutation_iterator"

[the remark this answers is not in the document]
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:463]: OK, yes.  To be clear, I just mean that when we merge, lets keep 2 final commits (after squash) from this PR, one for the bugfix and one for the rest.

[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:469]: yes, sure, I will keep 2 final commits.

**Review by danhoeflinger — APPROVED** [review]
LGTM
